### PR TITLE
Provide a custom diagnostic for conformance to NSObjectProtocol

### PR DESCRIPTION
**Pull request by DougGregor**
Swift classes cannot meaningfully conform to NSObjectProtocol.
Inheriting from NSObject is the appropriate fix, so suggest that.
Fixes rdar://problem/32543753.

The diagnostic now looks like this:

```
proto.swift:3:7: error: cannot declare conformance to 'NSObjectProtocol' in Swift; 'Foo' should inherit 'NSObject' instead
class Foo: NSObjectProtocol { }
      ^    ~~~~~~~~~~~~~~~~
           NSObject
```

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckProtocol.cpp [hunk 1]
```diff
@@ -2748,6 +2757,11 @@ diagnoseMissingWitnesses(MissingWitnessDiagnosisKind Kind) {
     auto &SM = DC->getASTContext().SourceMgr;
     auto FixitBufferId = SM.findBufferContainingLoc(FixitLocation);
     for (auto VD : MissingWitnesses) {
+      // Don't ever emit a diagnostic for a requirement in the NSObject
+      // protocol. They'renot implementable.
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: typo: "they're not"

**Review by hborla — APPROVED** [review]

File: lib/Sema/TypeCheckProtocol.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/Sema/TypeCheckProtocol.cpp]: Fixed now, thanks!

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: lib/Sema/TypeCheckProtocol.cpp [hunk 2]
```diff
@@ -4052,6 +4067,45 @@ static void diagnoseConformanceFailure(Type T,
     }
   }
 
+  // One cannot meaningfully declare conformance to the NSObject protocol
+  // in Swift. Suggest inheritance from NSObject instead.
+  if (isNSObjectProtocol(Proto)) {
+    if (T->getClassOrBoundGenericClass()) {
```
Comment (brentdax) [inline lib/Sema/TypeCheckProtocol.cpp]: I'm not familiar with the surrounding code—does this call return the same class you `dyn_cast<ClassDecl>()` to get later?

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Review by beccadax — APPROVED** [review]
One thing you might tweak, but otherwise looks alright!

**Pull request merged**